### PR TITLE
feat: add zls support

### PR DESCRIPTION
**Pull request by jinzhongjia**
## Features

- Code logic organization and optimization
- Add new command `remove`
- Destroyed the original download prompt message

Since the original download information is destroyed, it is not clear what kind of interaction should be better.

### CODE REVIEW

File: src/tools.zig [hunk 1]
```diff
@@ -45,3 +49,40 @@ pub fn get_zvm_path_segment(tmp_allocator: std.mem.Allocator, segment: []const u
         &[_][]const u8{ get_home(), ".zm", segment },
     );
 }
+
+/// for verify hash
+pub fn verify_hash(computed_hash: [32]u8, actual_hash_string: []const u8) bool {
```
[the remark this answers is not in the document]
Comment (jinzhongjia) [reply src/tools.zig:54]: Hash verification is just a tool function. It is used in zig download and zls download.
I think tucking it into tool.zig is a better division
More intuitive to use

**Comment by jinzhongjia** [comment]
Sorry, I made a lot of changes to the source code, but I think the changes were worth it and the readability should be improved by now

File: src/command.zig [hunk 2]
```diff
@@ -35,119 +39,227 @@ const CommandOption = struct {
...
+            is_zls = false;
+        } else if (tools.eql_str(scmd, "zls")) {
+            is_zls = true;
+        } else {
+            std.debug.print("Unknown subcommand '{s}'. Use 'remove zig <version>' or 'remove zls <version>'.\n", .{scmd});
+            return;
+        }
+
+        const version = param orelse {
+            std.debug.print("Please specify a version to use: 'remove zig/zls <version>'.\n", .{});
```
Comment (hendriknielaender) [inline src/command.zig:251]:
```suggestion
            std.debug.print("Please specify a version: 'remove zig <version>' or 'remove zls <version>'.\n", .{});
```

File: src/command.zig [hunk 3]
```diff
@@ -35,119 +39,227 @@ const CommandOption = struct {
...
+
+                var zls_meta = try meta.Zls.init(res, allocator);
+                defer zls_meta.deinit();
+
+                const version_list = try zls_meta.get_version_list(allocator);
+                break :blk version_list;
+            } else
+            // when not zig
+            if (!tools.eql_str(p, "zig")) {
+                std.debug.print("Error param, you can specify zig or zls\n", .{});
```
Comment (hendriknielaender) [inline src/command.zig:156]: On a general note, shouldn't we use `std.log` instead? We can tackle this in another PR.

File: src/command.zig [hunk 4]
```diff
@@ -35,119 +39,227 @@ const CommandOption = struct {
...
 
         const args = params[1..];
 
-        for (args, 0..) |arg, index| {
-            for (command_opts) |opt| {
-                const is_eql_short_handle = if (opt.short_handle) |short_handle|
-                    std.mem.eql(u8, arg, short_handle)
-                else
-                    false;
+        // for (args, 0..) |arg, index| {
```
Comment (hendriknielaender) [inline src/command.zig:56]: Leftover?

File: src/command.zig [hunk 5]
```diff
@@ -1,18 +1,22 @@
...
+const config = @import("config.zig");
 
 // Command types
 pub const Command = enum {
     List,
     Install,
     Use,
-    Default,
+    Remove,
+    // Default,
```
Comment (hendriknielaender) [inline src/command.zig:19]: Since we are not supporting Default so far, we could also remove it completely.

File: src/alias.zig [hunk 6]
```diff
@@ -1,145 +1,97 @@
...
-    try child_process.spawn();
+    const actual_version = try tools.get_current_version(allocator, true);
+    defer allocator.free(actual_version);
 
-    if (child_process.stdout) |stdout| {
-        const version = try stdout.reader().readUntilDelimiterOrEofAlloc(allocator, '\n', 100) orelse return error.EmptyVersion;
-        assert(version.len > 0);
-        return version;
+    if (!std.mem.eql(u8, expected_version, actual_version)) {
+        std.debug.print("Expected Zls version {s}, but currently using {s}. Please check.\n", .{ expected_version, actual_version });
```
Comment (hendriknielaender) [inline src/alias.zig:93]:
```suggestion
        std.debug.print("Expected zls version {s}, but currently using {s}. Please check.\n", .{ expected_version, actual_version });
```

File: src/alias.zig [hunk 7]
```diff
@@ -1,145 +1,97 @@
...
+    } else {
+        std.debug.print("Now using Zls version {s}\n", .{expected_version});
```
Comment (hendriknielaender) [inline src/alias.zig:95]:
```suggestion
        std.debug.print("Now using zls version {s}\n", .{expected_version});
```

File: src/alias.zig [hunk 8]
```diff
@@ -1,145 +1,97 @@
...
+                try tools.get_zvm_zig_version(arena_allocator),
+            version,
+        },
+    );
+
+    std.fs.accessAbsolute(version_path, .{}) catch |err| {
+        if (err != error.FileNotFound)
+            return err;
+
+        std.debug.print("zig {s} is not installed, please install it!\n", .{version});
```
Comment (hendriknielaender) [inline src/alias.zig:35]:
```suggestion
        std.debug.print("zig version {s} is not installed. Please install it before proceeding.\n", .{version});
```

**Review by hendriknielaender — COMMENTED** [review]
Great effort on the recent updates! The improvements are appreciated. It's exciting to see that we'll soon support zls as well, which will significantly enhance zvm. Keep up the excellent work!

**Comment by jinzhongjia** [comment]
Yes, actually I recommend not using `debug.print`, because it will write directly to stdout
Maybe we can override the default log function. The zig standard library provides a better overwriting method.
tools.zig has been split into source files under `util`

**Comment by jinzhongjia** [comment]
There is no problem when testing under Linux. I have not tested on Windows.

**Comment by hendriknielaender** [comment]
> There is no problem when testing under Linux. I have not tested on Windows.

Alright, I will test it for Windows and macOS

**Comment by jinzhongjia** [comment]
If the tests are OK, I think it can be merged, but it should not be released immediately because the interaction is broken.

I will try to add local compilation support for zzls and master update support for zig and zls later

File: src/command.zig [hunk 9]
```diff
@@ -35,119 +41,231 @@ const CommandOption = struct {
...
-    std.fs.accessAbsolute(current_zig_path, .{}) catch |err| {
+    const current_path = try std.fs.path.join(
+        allocator,
+        &.{ current, if (is_zls) "zls" else "zig" },
+    );
+
+    std.fs.accessAbsolute(current_path, .{}) catch |err| {
         if (err == std.fs.Dir.AccessError.FileNotFound) {
-            std.debug.print("Zig has not been installed yet, please install zig with zvm!\n", .{});
+            std.debug.print("{s} has not been installed yet, please install it fist!\n", .{if (is_zls) "Zls" else "Zig"});
```
Comment (hendriknielaender) [inline src/command.zig:134]:
```suggestion
            std.debug.print("{s} has not been installed yet, please install it fist!\n", .{if (is_zls) "zls" else "Zig"});
```

File: src/command.zig [hunk 10]
```diff
@@ -166,16 +284,19 @@ fn display_help() !void {
         \\    zvm <command> [args]
         \\
         \\Commands:
-        \\    ls, list       List the versions of Zig available to zvm.
-        \\    i, install     Install the specified version of Zig.
-        \\    use            Use the specified version of Zig.
-        \\    --version      Display the currently active Zig version.
-        \\    --default      Set a specified Zig version as the default for new shells.
+        \\    ls, list       List the versions of Zig or zls available to zvm.
+        \\    i, install     Install the specified version of Zig or Zls.
```
Comment (hendriknielaender) [inline src/command.zig:288]:
```suggestion
        \\    i, install     Install the specified version of Zig or zls.
```

File: src/command.zig [hunk 11]
```diff
@@ -166,16 +284,19 @@ fn display_help() !void {
...
+        \\    use            Use the specified version of Zig Zls.
```
Comment (hendriknielaender) [inline src/command.zig:289]:
```suggestion
        \\    use            Use the specified version of zig or zls.
```

File: src/command.zig [hunk 12]
```diff
@@ -166,16 +284,19 @@ fn display_help() !void {
...
+        \\    remove         Remove the specified version of Zig or Zls
```
Comment (hendriknielaender) [inline src/command.zig:290]:
```suggestion
        \\    remove         Remove the specified version of Zig or zls
```

**Comment by jinzhongjia** [comment]
Hey man, I don't know if you are willing to accept this PR now, If you are not willing to accept the PR, you can close it

**Comment by hendriknielaender** [comment]
> Hey man, I don't know if you are willing to accept this PR now, If you are not willing to accept the PR, you can close it

I'm willing to accept this, but the minimum requirement for me would be that this is tested on windows as well. Last week I didn't had time to verify your changes, feel free to manual test this.

**Comment by jinzhongjia** [comment]
now, windows test ok

Due to the introduction of new code, macos and linux platforms need to be retested

**Comment by jinzhongjia** [comment]
I will try linux testing later

**Comment by jinzhongjia** [comment]
Linux tests OK.

**Pull request merged**